Add Extension Headers to PUBLISH, SUBSCRIBE_OK, and FETCH_OK (#1374)
moq-wg/moq-transport: draft-ietf-moq-transport.md

3. Use TRACK_STATUS or similar mechanism to query the previous state to
   determine the largest published Group ID.

## Extension Headers {#extension-headers}

Tracks and Objects can have additional relay-visible fields, known as Extensions,
which do not require negotiation, and can be used to alter
MoQT Object distribution.

Extension Headers are defined in {{moqt-extension-headers}} as well as external
specifications and are registered in an IANA table {{iana}}. These
specifications define the type and value of the header, along with any rules
concerning processing, modification, caching and forwarding. A relay which is

> coded to implement these rules is said to "support" the extension.

Reviewer: this is bit inconsistent with relay handling section, which has MUST forward in all the cases. The rules in this paragraph applies to object extensions in general.

3. Use TRACK_STATUS or similar mechanism to query the previous state to
   determine the largest published Group ID.

## Extension Headers {#extension-headers}

> 

Reviewer: another simplicity is , may be remove the word Headers and call it Extensions. Rename this section to ```Track/Object Extensions```

Tracks and Objects can have additional relay-visible metadata called Extension
Headers.

Extension Headers are visible to relays, do not require negotiation, and can
be used to alter MoQT Object distribution.

Extension Headers are defined in {{moqt-extension-headers}} as well as external
specifications and are registered in an IANA table {{iana}}. These
specifications define the type and value of the header, along with any rules
concerning processing, modification, caching and forwarding. A relay which is
coded to implement these rules is said to "support" the extension.

If unsupported by the relay, Extension Headers MUST NOT be modified, MUST be
cached as part of the Track or Object and MUST be forwarded by relays.

If supported by the relay and subject to the processing rules specified in the
definition of the extension, Extension Headers MAY be modified, added, removed,

> and/or cached by relays.

Reviewer 2: Do we need to mention IMMUTABLE?

3. Use TRACK_STATUS or similar mechanism to query the previous state to
   determine the largest published Group ID.

> ## Extension Headers {#extension-headers}

Reviewer: ```suggestion
## MOQ Extensions {#extension-headers}
```

Tracks and Objects can have additional relay-visible fields, known as Extension
Headers, which do not require negotiation, and can be used to alter
MoQT Object distribution.

Extension Headers are defined in {{moqt-extension-headers}} as well as external
specifications and are registered in an IANA table {{iana}}. These
specifications define the type and value of the header, along with any rules
concerning processing, modification, caching and forwarding.

If unsupported by the relay, Extension Headers MUST NOT be modified, MUST be
cached as part of the Track or Object and MUST be forwarded by relays.

If supported by the relay and subject to the processing rules specified in the

> definition of the extension, Extension Headers MAY be modified, added, removed,

Reviewer: ```suggestion
definition of the extension, Extensions MAY be modified, added, removed,
```

3. Use TRACK_STATUS or similar mechanism to query the previous state to
   determine the largest published Group ID.

## Extension Headers {#extension-headers}

Tracks and Objects can have additional relay-visible fields, known as Extensions,

> which do not require negotiation, and can be used to alter

Author: ```suggestion
Tracks and Objects can have additional relay-visible fields, known as Extension
Headers, which do not require negotiation, and can be used to alter
```

MoQT Object distribution.

Extension Headers are defined in {{moqt-extension-headers}} as well as external
specifications and are registered in an IANA table {{iana}}. These
specifications define the type and value of the header, along with any rules
concerning processing, modification, caching and forwarding. A relay which is

> coded to implement these rules is said to "support" the extension.

Author: I'm removing this sentence because it's non-normative and inconsistent as you said.

```suggestion
concerning processing, modification, caching and forwarding.
```

1. Object Status can transition from any status to Object Does Not Exist in
   cases where the object is no longer available.  Transitions between Normal,
   End of Group and End of Track are invalid.

> 3. Object Extension Headers can be added, removed or updated, subject

Reviewer 2: ```suggestion
2. Object Extension Headers can be added, removed or updated, subject
```

published namespaces and subscriptions to the old relay can be withdrawn or
terminated.

## Relay Track Handling

Relays that receive Track Extension Headers MUST include them in any PUBLISH,
SUBSCRIBE_OK, TRACK_STATUS_OK or FETCH_OK sent for that Track. Relays
MUST NOT modify or remove Extension Headers unless the relay supports the

> Extension (see {{extension-headers}}).

Reviewer: ```suggestion
SUBSCRIBE_OK, TRACK_STATUS_OK or FETCH_OK sent for that Track. Relays
MUST follow the rules defined in (see {{extension-headers}}).
```

published namespaces and subscriptions to the old relay can be withdrawn or
terminated.

## Relay Track Handling

A relay MUST include Track Extension Headers when sending any PUBLISH,
SUBSCRIBE_OK, TRACK_STATUS_OK or FETCH_OK, unless allowed by the extension's

> specification (see {{extension-headers}}).

Author: I'm happy with this change.

* Parameters: The parameters are defined in {{version-specific-params}}.

> * Track Extensions : A sequence of Extension Headers. See {{extension-headers}}.

Reviewer: ```suggestion
* Track Extensions : A sequence of Track Extensions. See {{extension-headers}}.
```

#### Object Extension Header {#object-extensions}

Any Object with status Normal can have extension headers ({{extension-headers}}).
If an endpoint receives extension headers on Objects with status that is
not Normal, it MUST close the session with a `PROTOCOL_VIOLATION`.

> Object Extension Headers are visible to relays and are intented to be relevant

Reviewer 3: ```suggestion
Object Extension Headers are visible to relays and are intended to be relevant

be closed with a `PROTOCOL_VIOLATION` ({{session-termination}}).
Any object with a status code other than zero MUST have an empty payload.

#### Object Extension Headers {#object-extensions}

Any Object with status Normal can have extension headers ({{extension-headers}}).
If an endpoint receives extension headers on Objects with status that is
not Normal, it MUST close the session with a `PROTOCOL_VIOLATION`.

> Object Extension Headers are visible to relays and are intended to be relevant

Reviewer: ```suggestion
Object Extensions are visible to relays and are intended to be relevant
```

to MOQT Object distribution. Any Object metadata never intended to be accessed
by the transport or Relays SHOULD be serialized as part of the Object payload
and not as an extension header.

> Object Extension Headers are serialized as a length in bytes followed by

Reviewer: ```suggestion
Object Extensions are serialized as a length in bytes followed by
```

Key-Value-Pairs (see {{moq-key-value-pair}}).

~~~
Extensions {
  Extension Headers Length (i),
  Extension Headers (..),
}
~~~

Object Extension Header types are registered in the IANA table

> 'MOQ Extension Headers'. See {{iana}}.

Reviewer: ```suggestion
'MOQ Object Extensions '. See {{iana}}.
```

~~~

> # Extension Headers {#moqt-extension-headers}

Reviewer: ```suggestion
# MOQ Extensions  {#moqt-extension-headers}
```

An Object MUST NOT contain more than one instance of this extension header.

> ## Prior Group ID Gap

Reviewer 2: Do we need to say "Object only" extension? Same for Prior Object ID Gap?
